vm/systemSmartContracts/governance.go

```diff
@@ -288,521 +499,599 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+// closeProposal generates and saves end results for a proposal
+func (g *governanceContract) closeProposal(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if args.CallValue.Cmp(zero) != 0 {
+		g.eei.AddReturnMessage("closeProposal callValue expected to be 0")
+		return vmcommon.UserError
 	}
 	if !g.isWhiteListed(args.CallerAddr) {
-		g.eei.AddReturnMessage("called address is not whiteListed")
+		g.eei.AddReturnMessage("caller is not whitelisted")
 		return vmcommon.UserError
 	}
-	gitHubCommit := args.Arguments[2]
-	if len(gitHubCommit) != githubCommitLength {
-		g.eei.AddReturnMessage(fmt.Sprintf("invalid github commit length, wanted exactly %d", githubCommitLength))
+	if len(args.Arguments) != 1 {
+		g.eei.AddReturnMessage("invalid number of arguments expected 1")
 		return vmcommon.UserError
 	}
-	if g.proposalExists(gitHubCommit) {
-		g.eei.AddReturnMessage("proposal already exists")
+	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.CloseProposal)
+	if err != nil {
+		g.eei.AddReturnMessage("not enough gas")
+		return vmcommon.OutOfGas
+	}
+
+	proposal := args.Arguments[0]
+	generalProposal, err := g.getGeneralProposal(proposal)
+	if err != nil {
+		g.eei.AddReturnMessage("getGeneralProposal error " + err.Error())
 		return vmcommon.UserError
 	}
+	if generalProposal.Closed {
+		g.eei.AddReturnMessage("proposal is already closed, do nothing")
+		return vmcommon.Ok
+	}
 
-	key := append([]byte(hardForkPrefix), gitHubCommit...)
-	marshaledData := g.eei.GetStorage(key)
-	if len(marshaledData) != 0 {
-		g.eei.AddReturnMessage("hardFork proposal already exists")
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+	if currentNonce < generalProposal.EndVoteNonce {
+		g.eei.AddReturnMessage(fmt.Sprintf("proposal can be closed only after nonce %d", generalProposal.EndVoteNonce))
 		return vmcommon.UserError
 	}
 
-	startVoteNonce, endVoteNonce, err := g.startEndNonceFromArguments(args.Arguments[3], args.Arguments[4])
+	generalProposal.Closed = true
+	err = g.computeEndResults(generalProposal)
 	if err != nil {
-		g.eei.AddReturnMessage("invalid start/end vote nonce" + err.Error())
+		g.eei.AddReturnMessage("computeEndResults error" + err.Error())
 		return vmcommon.UserError
 	}
 
-	bigIntEpochToHardFork, okConvert := big.NewInt(0).SetString(string(args.Arguments[0]), conversionBase)
-	if !okConvert || !bigIntEpochToHardFork.IsUint64() {
-		g.eei.AddReturnMessage("invalid argument for epoch")
+	err = g.saveGeneralProposal(proposal, generalProposal)
+	if err != nil {
+		g.eei.AddReturnMessage("saveGeneralProposal error" + err.Error())
 		return vmcommon.UserError
 	}
 
-	epochToHardFork := uint32(bigIntEpochToHardFork.Uint64())
-	currentEpoch := g.eei.BlockChainHook().CurrentEpoch()
-	if epochToHardFork < currentEpoch && currentEpoch-epochToHardFork < hardForkEpochGracePeriod {
-		g.eei.AddReturnMessage("invalid epoch to hardFork")
-		return vmcommon.UserError
+	for _, voter := range generalProposal.Votes {
```

Comment:
we have to test if one such call can take more than 1 second.

Reply:
Yes, it in not necessary to do this (these are also only the vote identifiers). We could remove the actual votes too, but if it takes too long we can leave them in the storage.

Reply:
make a separate function which does the cleanup and has a cost per delete operation - thus will clean only how much it can. We should be careful as well that the leafs do not get too big - thus generalProposal.Votes should have only a few thousand votes. Another idea is to make the keys for the VOTES deterministic - thus no need to keep them in a list.

Reply:
make a separate function which does the cleanup and has a cost per delete operation - thus will clean only how much it can. We should be careful as well that the leafs do not get too big - thus generalProposal.Votes should have only a few thousand votes. Another idea is to make the keys for the VOTES deterministic - thus no need to keep them in a list.